feat(Apollo): make use of generic type on MutationOptions in Apollo#mutate

Merott opened the pull request:
Without this the `MutationOptions` generic takes the default type of `{ [key: string]: any; }` which isn't very useful. This is needed by the `updateQueries` option to detect the right type for the `mutationResult` data object.

kamilkisiela commented:
Hi! I will update everything around Angular and TypeScript soon, then I'm going to merge this :)

kamilkisiela commented:
@Merott Hi! :) Could you rebase it?

Merott commented:
I just rebased, but in doing so I also removed the explicit types on the method, since TypeScript can infer them perfectly fine, and making them explicit makes human errors more likely.

The same could be done for `query`, `mutate` and `subscribe`. I've verified that on my machine but not committed. I'd be happy to commit and push if you want.
